### PR TITLE
Add aws support - Samuel's version without CSR

### PR DESCRIPTION
**Pull request by Bravo555**
## Proposed changes

This PR is basically #1460 but with resolved conflicts and without the commit adding CSR, as it was decided that eagerly generating the CSR and having it sit alongside the self-signed certificate would perhaps be confusing for users. Instead, the CSR will be looked at separately in another issue/PR, either being generated by invoking another command, or being used by AWS CLI to obtain a certificate signed by AWS IoT CA like in Bravo555/thin-edge.io@cbfeee784fac3d25998e125eea3c3ba215e9aba4.

The last thing to do before merging this is to revise the topic mappings in the AWS mapper.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

Relevant documentation:

- [MQTT topics](https://docs.aws.amazon.com/iot/latest/developerguide/topics.html)
- [Reserved topics](https://docs.aws.amazon.com/iot/latest/developerguide/reserved-topics.html)
- [Designing MQTT Topics for AWS IoT Core](https://docs.aws.amazon.com/whitepapers/latest/designing-mqtt-topics-aws-iot-core/designing-mqtt-topics-aws-iot-core.html)

### CODE REVIEW

File: crates/core/tedge/src/cli/connect/bridge_config_aws.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        let address = format!("{}:{}", connect_url.as_str(), mqtt_tls_port);
+        let user_name = remote_clientid.to_string();
+        let pub_msg_topic = format!(
+            "messages/# out 1 aws/ thinedge/devices/{}/",
+            remote_clientid
+        );
+        let sub_msg_topic = format!(
+            "messages/devicebound/# in 1 aws/ thinedge/devices/{}/",
+            remote_clientid
+        );
```
Comment (didier-wenzek) [inline crates/core/tedge/src/cli/connect/bridge_config_aws.rs:37]: One needs to change something here.

* It sounds unusual to have inbound messages as a subset of outbound messages.
*  There is risk of an infinite loop generated by the bridge. A message published on the device to `aws/messages/devicebound/foo` will be echoed on the device to `aws/messages/devicebound/foo`. 

Why not simply a specific suffix for each direction, say `outbound` and `inbound`:

```suggestion
        let pub_msg_topic = format!(
            "messages/outbound/# out 1 aws/ thinedge/devices/{}/",
            remote_clientid
        );
        let sub_msg_topic = format!(
            "messages/inbound/# in 1 aws/ thinedge/devices/{}/",
            remote_clientid
        );
```

File: docs/src/references/bridged-topics.md [hunk 2]
```diff
@@ -64,6 +64,17 @@ The Azure topics are prefixed by `az/`.
...
+ ## AWS MQTT Topics
+MQTT clients on Thin Edge device must use the below topics to communicate with the AWS cloud.
+The AWS topics are prefixed by `aws/`.
+
+ * `aws/messages/`  - Use this topic to send the messages from device to cloud.
+ The messages are forwarded to the AWS topic named `thinedge/devices/{device_id}/messages/`
+ where {device_id} is the Thin Edge device id.
+
+ * `aws/messages/devicebound/#` - Use this topic to subscribe for the messages that were sent from cloud to device.
+ Any message published by AWS on one the subtopics of `thinedge/devices/{device_id}/messages/devicebound/#`
```
Comment (didier-wenzek) [inline docs/src/references/bridged-topics.md:76]: This naming scheme needs to be revised. See my comment on the bridge configuration.

**Review by didier-wenzek — COMMENTED** [review]
With the exception of the bridge topics which needs to be discussed, this PR is ready to be merged. Thank you!

File: docs/src/tutorials/connect-aws.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+```
+
+You may notice that the issuer of this certificate is the device itself.
+This is a self-signed certificate.
+To use a certificate signed by your Certificate Authority,
+see the reference guide of [`tedge cert`](../references/tedge-cert.md).
+
+## Register the device on AWS IoT Hub
+
+For a device to be trusted by AWS, one needs to add the self-signed certificate or certificate signing request file to the AWS IoT Core (both are generated when using the `tedge cert create` command).
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply docs/src/tutorials/connect-aws.md:57]: Is it really hard to automate certificate upload? like `tedge cert upload c8y`?

**Review by PradeepKiruvale — COMMENTED** [review]
LGTM

File: docs/src/tutorials/connect-aws.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bravo555) [reply docs/src/tutorials/connect-aws.md:57]: > Is it really hard to automate certificate upload? like `tedge cert upload c8y`?

To do this, we would have to bring in AWS IoT, like in my PoC in the PR description, bringing in additional dependencies, or wrap up some custom solution with a lot of code.

File: crates/core/tedge/src/cli/connect/bridge_config_aws.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/tedge/src/cli/connect/bridge_config_aws.rs:37]: As proposed by @Bravo555, it makes sense to bridge the AWS reserved `shadow` topic

```
topic # both 1 aws/device $aws/things/{client_id}/shadow
```

About measurements, alarms, and events, I would not enforce those but let the users free to add its own suffix and use the same `td` prefix as used by AWS. For inbound messages, which are commands, I would use the `cmd` prefix, again to be consistent with the AWS examples.

```
topic td/# out 1 aws/ thinedge/{device-id}/",
topic cmd/# in 1 aws/ thinedge/{device-id}/
```

**Comment by Bravo555** [comment]
@didier-wenzek had to spend some time testing, but I think I finished with the topic mapping. Let me know if it looks any good, and if there's no issues I'll squash the commits before the final merge.

**Comment by didier-wenzek** [comment]
> @didier-wenzek had to spend some time testing, but I think I finished with the topic mapping. Let me know if it looks any good, and if there's no issues I'll squash the commits before the final merge.

The updated topic scheme is good.

I will approved the PR once the checks clean-up.

**Review by didier-wenzek — APPROVED** [review]
Approved

**Review by didier-wenzek — APPROVED** [review]
Approved.

There is a failing system test - however not introduced by this PR and tracked here https://github.com/thin-edge/thin-edge.io/issues/1731

**Pull request merged**